src/plots/Histogram.tsx

```diff
@@ -177,31 +170,153 @@ export default function Histogram({
           marker: {
             ...(series.color ? { color: series.color } : {}),
-            opacity: perBarOpacity,
           },
           offset: 0,
           width: binWidths,
+          selected: {
+            marker: {
+              opacity: 1,
+            },
+          },
+          unselected: {
+            // switch off fading of unselected bars while selecting
+            marker: {
+              opacity: 1,
+            },
+          },
         };
       }),
     [data, orientation, calculatedBarOpacity, selectedRange]
   );
 
-  const handleSelectedRange = useCallback(
+  /**
+   * calculate midpoints of a unique set of bins
+   */
+  const binSummaries: BinSummary[] = useMemo(() => {
+    const allBins: HistogramBin[] = data.series
+      .map((series) => [series.bins])
+      .flat(2);
+
+    const seenLabels: string[] = [];
+    const uniqueBins = allBins.filter((bin) => {
+      const seenBefore = seenLabels.indexOf(bin.binLabel);
+      seenLabels.push(bin.binLabel);
+      return seenBefore < 0;
+    });
+    // sort them on binStart just in case
+    uniqueBins.sort((bina, binb) =>
+      bina.binStart > binb.binStart
+        ? 1
+        : bina.binStart === binb.binStart
+        ? 0
+        : -1
+    );
+
+    // return the list of summaries - note the binMiddle prop
+    return uniqueBins.map((bin) => ({
+      binStart: bin.binStart,
+      binEnd: bin.binEnd,
+      binMiddle:
+        data.valueType === 'date'
+          ? DateMath.add(
+              bin.binStart as Date,
+              DateMath.diff(
+                bin.binStart as Date,
+                bin.binEnd as Date,
+                'seconds',
+                false
+              ) * 500,
+              'milliseconds'
+            )
+          : ((bin.binStart as number) + (bin.binEnd as number)) / 2.0,
+    }));
+  }, [data.series, data.valueType]);
+
+  // local state for range **while selecting** graphically
+  const [selectingRange, setSelectingRange] = useState<NumberOrDateRange>();
+
+  const handleSelectingRange = useCallback(
     (object: any) => {
       if (object && object.range) {
         const [val1, val2] =
           orientation === 'vertical' ? object.range.x : object.range.y;
         const [min, max] = val1 > val2 ? [val2, val1] : [val1, val2];
         // TO DO: think about time zones?
-        onSelectedRangeChange({
+        const rawRange: NumberOrDateRange = {
           min: data.valueType === 'date' ? new Date(min) : min,
           max: data.valueType === 'date' ? new Date(max) : max,
-        } as NumberOrDateRange);
+        };
+
+        // now snap to bin boundaries using same logic that Plotly uses
+        // (dragging range past middle of bin selects it)
+        const leftBin = binSummaries.find(
+          (bin) => rawRange.min < bin.binMiddle
+        );
+        const rightBin = binSummaries
+          .slice()
+          .reverse()
+          .find((bin) => rawRange.max > bin.binMiddle);
```

Comment:
I'm not convinced that including bins based on whether their middle is selected is intuitive. I tried out including a bin if any of its range is selected, and that feels far better to me. Here's that version for reference if someone else wants to try it:

```suggestion
        const leftBin = binSummaries.find(
          (bin) => rawRange.min < bin.binEnd && rawRange.min > bin.binStart
        );
        const rightBin = binSummaries
          .slice()
          .reverse()
          .find((bin) => rawRange.max > bin.binStart && rawRange.max < bin.binEnd);
```

Reply:
Interesting, thanks Connor.

BTW, the middle-based behaviour is what Plotly does, and what we have tried to emulate.

I tested your version, and I think it tends to include an extra bin to the left of where you start dragging (left to right in my case). But maybe I got used to the Plotly behaviour...

There's also an issue with drag-selecting to the L or R extremes of the plot.  The selection is lost.  It can be fixed with the >= and <= added as below, but only for the numeric histogram (test via /story/plots-histogram--range-selection) - the date-based one still has problems (/story/plots-histogram--date-range-selection) - but maybe that can be fixed - it could be 00:00 vs 23:59:59.999 issue.

```
        const leftBin = binSummaries.find(
          (bin) => rawRange.min < bin.binEnd && rawRange.min >= bin.binStart
        );
        const rightBin = binSummaries
          .slice()
          .reverse()
          .find((bin) => rawRange.max > bin.binStart && rawRange.max <= bin.binEnd);

```

Reply:
Good points! Dave agrees with you also, so I'll happily consent 👍